ci-operator/README.md

```diff
@@ -0,0 +1,154 @@
+
+The first step is writing a configuration file for `ci-operator` for the
+repository.  Many projects already have one, which takes care of most of the CI
+process: downloading the source, building binaries, building RPMs, creating
+images, executing unit tests, etc.  These can be built upon for e2e test,
+requiring little or no modification.
+
+After that, to add an e2e test:
+
+1. Determine the pre-requisites for the test.  Practically, this means choosing
+   the template that is going to be used by `ci-operator`.  This directory
+   already contains templates for the common cases, see
+   [Using a template](#using-a-template) below.
+2. Determine and configure the template's inputs.  This is specific to each
+   template and should be documented in its parameters.  The e2e test might
+   need minor modifications to fit the environment created by the template.
+3. Add one or more prow jobs to prow's configuration file with the information
+   gathered in the previous steps.
+
+
+# Provisioning a cluster
+
+Contrary to other types of tests, e2e tests usually require a cluster, not just
+a single container.  While there aren't yet native primitives in `ci-operator`
+for cluster provisioning, it provides one open-ended feature that can be
+leveraged to accomplish that: template steps.
+
+Template steps allow the creation of arbitrary objects in the cluster where the
+CI pipeline is executed.  This is used to start a pod that will then provision
+a separate cluster for the tests.  This directory already contains a few
+templates that can be used either directly or (rarely necessary, in practice)
+as a reference:
+
+- `cluster-launch-e2e.yaml`: launches a cluster in GCP using openshift-ansible
+  and runs Origin e2e tests on it, parameterized by test focus.
+- `cluster-launch-installer-e2e.yaml`:
+- `cluster-launch-src.yaml`: launches a cluster in GCP using openshift-ansible
+  and runs a script from the repository being tested with the resulting
+  `$KUBECONFIG`, parameterized by test script.
+- `master-sidecar.yaml`: spins up a simple openshift control plane as a sidecar
+  and waits for the `COMMAND` specified to the template to be executed, before
+  itself exiting. The test container is given access to the generated
+  configuration and the `admin.kubeconfig`.
+- `os.yaml`:
+
+
+# Using a template
+
+A template can be used as a stage in `ci-operator` with the `--template`
+argument.  The new template stage is added at runtime to those read from the
+configuration file.  The name of the template, or the name of the file without
+the extension, becomes the name of the stage, so it can be used as an argument
+to `--target`.
+
+When processing the template, `ci-operator` passes several parameters with
+information about the tests (e.g. the job name, which namespace it's operating
+into, URLs for images).  These parameters are described in its usage message.
+
+Logs from pods that fail will be fetched and stored if collection of artifacts
```

Comment:
"Test artifacts (such as logs or other arbitrary data) can be automatically retrieved and made available in <whatever system is responsible for this>. To enable artifact collection, specify the `--artifacts-dir=<some path>` option) and write your test to place the artifacts into this location."

Really minor pedant nit: a "flag" is a binary on/off option, so technically `--artifacts-dir=<some string>` isn't a flag, it's an option.

Reply:
Github ate part of my sentence... I meant to write "made available in [whatever Prow system is used to view test artifacts]".

Reply:
I changed the sentence to explain what is done specifically for templates when
artifact collection is enabled, with a link to the documentation (which AFAIK
doesn't yet exist) for more information.  There are many aspects to artifact
collection and I don't think this is the place to document them.

Let me know if that's not enough.

Reply:
Also: fixed other flag/option mistakes (I agree 100% on the distinction, no
idea how so many got through).